### PR TITLE
PublicKey: add read_from method to be symmetric with write_into

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

File: src/util/key.rs
Comment (stevenroose) [inline src/util/key.rs:124]: indentation

Comment (apoelstra) [reply src/util/key.rs]: fixed. I need to find whatever broken vim setting is making it do that..

File: src/util/key.rs
Comment (sgeisler) [inline src/util/key.rs:528]: nit: would be more elegant as a iterator chain `(0..N_KEYS).map(…).collect::<Vec<_>>()`, but it's only a test.

**Review by sgeisler — DISMISSED** [review]
ACK 79064beec74765c7dde8d6af384449ec90330cea

**Review by sgeisler — APPROVED** [review]
ACK 8881e494f225431d3ea8eb1729234d002686c017

File: src/util/key.rs
Comment (Kixunil) [inline src/util/key.rs:123]: I believe that `let bytes = if ...;` expression is considered idiomatic.

Or perhaps more concisely:

```rust
let mut bytes = [0; 65];
// using ..1 instead of 0..1 avoids needless bound check, so I tend to prefer it
reader.read_exact(&mut bytes[..1])?;
// shadowing is considered fine here
let bytes = if bytes[0] < 4 {
    &mut bytes[..33]
} else {
    &mut bytes[..65]
}
};

reader.read_exact(&mut bytes[1..])?;
Self::from_slice(bytes).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
```

BTW since this reads a single byte at the beginning, perhaps add a doc comment recommending to use buffered reader.

Comment (sgeisler) [reply src/util/key.rs:123]: +1 for the doc comment.

Comment (apoelstra) [reply src/util/key.rs:123]: If the compiler generates different code for `0..1` than `..1` then I'll eat my hat :)

Everything else I agree with.

Comment (apoelstra) [reply src/util/key.rs]: pushed; took your code pretty-much verbatim and added a doccomment

**Review by sanket1729 — COMMENTED** [review]
Should we also try to address #520 here? Or maybe that can be after release.

File: src/util/key.rs
Comment (sanket1729) [inline src/util/key.rs:504]: What do people think about making this pub? I think all repositories that depend on rust-bitcoin have some function that does exactly this.

File: src/util/key.rs
Comment (Kixunil) [reply src/util/key.rs]: > If the compiler generates different code for 0..1 than ..1 then I'll eat my hat :)

Yeah, I believe it should be same in release mode. It's just a random quirk in my brain to prefer it. :)

File: src/util/key.rs
Comment (Kixunil) [reply src/util/key.rs:504]: Please NO! This is **not CSRNG**!

If people want it public, then please call it `not_crypto_safe_random_key_i_swear_to_use_it_for_testing_only`

Comment (apoelstra) [reply src/util/key.rs:504]: For what it's worth, nobody knows the discrete logs of the generated public keys, so it's probably really hard to actually cause a crypto break with this function.

But I'm curious what the purpose of it would be outside of unit tests :).

Comment (sanket1729) [reply src/util/key.rs:504]: Agree that is not CSRNG and it is not a good idea even if there is no secret key with it. It is already under mod tests, if we want it pub we can name it something else.

**Review by sgeisler — APPROVED** [review]
ACK 92000bf51bbeeb2e5d2b743a6566176bb07a384d

**Comment by sgeisler** [comment]
@sanket1729 I'd rather not add scope creep. If we get a second review soon I'd just merge, if not feel free to add a commit to address #520.

**Comment by sanket1729** [comment]
ACk 92000bf51bbeeb2e5d2b743a6566176bb07a384d. Let's get a release soon :)

**Comment by apoelstra** [comment]
@sanket1729 can you re-ack?

**Review by sanket1729 — APPROVED** [review]
ReAck- 92000bf51bbeeb2e5d2b743a6566176bb07a384d

**Pull request merged**
